graph-pyramid-builder/src/main.py

```diff
@@ -0,0 +1,625 @@
+            bins[feat1,feat2,rows[0],cols[0]] = ind2[0] + 1
+            bins[feat1,feat2,rows[1:],cols[1:]] = counts
+            linear_index.append([feat1,feat2])
+    return bins, bin_stats, linear_index
+
+""" 2. Plot Generation """
+
+# Tick formatting to mimick D3
+def format_ticks(fmin,fmax,nticks):
+    """ Generate tick labels
+
+    Polus Plots uses D3 to generate the plots. This function tries to mimic
+    the formatting of tick labels. Tick labels have a fixed width, and in
+    place of using scientific notation a scale prefix is appended to the end
+    of the number. See _prefix comments to see the suffixes that are used.
+
+    Numbers that are larger or smaller than 10**24 or 10**-24 respectively
+    are not handled and may throw an error. Values outside of this range
+    do not currently have an agreed upon prefix in the measurement science
+    community.
+
+    Inputs:
+        fmin - the minimum tick value
+        fmax - the maximum tick value
+        nticks - the number of ticks
+
+    Outputs:
+        fticks - a list of strings containing formatted tick labels
+
+    """
+    _prefix = {-24: 'y',  # yocto
+               -21: 'z',  # zepto
+               -18: 'a',  # atto
+               -15: 'f',  # femto
+               -12: 'p',  # pico
+                -9: 'n',  # nano
+                -6: 'u',  # micro
+                -3: 'm',  # mili
+                 0: ' ',
+                 3: 'k',  # kilo
+                 6: 'M',  # mega
+                 9: 'G',  # giga
+                12: 'T',  # tera
+                15: 'P',  # peta
+                18: 'E',  # exa
+                21: 'Z',  # zetta
+                24: 'Y',  # yotta
+                }
+    out = [t for t in np.arange(fmin,fmax,(fmax-fmin)/(nticks-1))]
+    out.append(fmax)
+    scale = np.log10(np.abs(out))
+    scale[np.isinf(scale)] = 0
+    scale_order = np.int8(3*np.sign(scale)*np.int8(scale/3))
+    fticks = []
+    for i in range(nticks):
+        fticks.append('{:{width}.{prec}f}'.format(out[i]/10**scale_order[i],
+                                                  width=3,
+                                                  prec=2-np.mod(np.int8(scale[i]),3)) + _prefix[scale_order[i]])
+    return fticks
+
+def get_cmap():
+    # Create a custom colormap to mimick Polus Plots
+
+    cmap_values = [[1.0,1.0,1.0,1.0]]
+    cmap_values.extend([[r/255,g/255,b/255,1] for r,g,b in zip(np.arange(0,255,2),
+                                                            np.arange(153,255+1/128,102/126),
+                                                            np.arange(34+1/128,0,-34/126))])
+    cmap_values.extend([[r/255,g/255,b/255,1] for r,g,b in zip(np.arange(255,136-1/128,-119/127),
+                                                            np.arange(255,0,-2),
+                                                            np.arange(0,68+1/128,68/127))])
+    cmap = ListedColormap(cmap_values)
+
+    return cmap
+
+def gen_plot(col1,
```

Comment:
I feel as though if you did here, what you did in the z-slice stack plugin, you could thread out the graph generation to N threads as well, speeding up this quite a bit.

Reply:
The plugin uses multiple threads in the pyramid building process (up to 4 depending on the number of cores). Generating processes for individual graphs would actually slow the plugin down because of the overhead required to spawn a process. So, what I do is spawn a process that builds a subpyramid 5 layers higher than the maximum resolution so each process will generate 32 graphs and create a minipyramid out of each one.

One thing I could do is generalize this process to every set of layers to the pyramid. So I could spawn all processes that would generate all subpyramids at 5 levels higher than the maximum resolution, and then spawn a separate set of processes to generate subpyramids for 5-10, etc. I haven't done this yet because it would require me to write a pyramid loading function to load specific layers of a pyramid. This would likely be faster for nodes with more than 4 cores, but since most of the nodes we work with in the cloud aren't that powerful I haven't seen the need to invest the time into it.